core/embed/rust/src/protobuf/encode.rs

```diff
@@ -0,0 +1,237 @@
+                    // fields also as a list-of-values. Serialize them as if
+                    // concatenated.
+
+                    let mut iter_buf = IterBuf::new();
+
+                    // Serialize the total length of the buffer.
+                    let mut len = 0;
+                    let iter = Iter::try_from_obj_with_buf(value, &mut iter_buf)?;
+                    for value in iter {
+                        let buffer = Buffer::try_from(value)?;
+                        len += buffer.len();
+                    }
+                    stream.write_uvarint(len as u64)?;
+
+                    // Serialize the buffers one-by-one.
+                    let iter = Iter::try_from_obj_with_buf(value, &mut iter_buf)?;
+                    for value in iter {
+                        let buffer = Buffer::try_from(value)?;
+                        stream.write(&buffer)?;
+                    }
+                } else {
+                    // Single length-delimited field.
+                    let buffer = Buffer::try_from(value)?;
+                    stream.write_uvarint(buffer.len() as u64)?;
+                    stream.write(&buffer)?;
+                }
+            }
+            FieldType::Msg(msg_type) => {
+                // Calculate the message size by encoding it through `CountingWriter`.
+                let counter = &mut CounterStream { len: 0 };
+                self.encode_message(counter, &msg_type, value)?;
+
+                // Encode the message as length-delimited bytes.
+                stream.write_uvarint(counter.len as u64)?;
+                self.encode_message(stream, &msg_type, value)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub trait OutputStream {
+    fn write(&mut self, buf: &[u8]) -> Result<(), Error>;
+    fn write_byte(&mut self, val: u8) -> Result<(), Error>;
+
+    fn write_uvarint(&mut self, mut num: u64) -> Result<(), Error> {
+        loop {
+            let shifted = num >> 7;
+            let byte = (num & 0x7F) as u8;
+            if shifted != 0 {
+                num = shifted;
+                self.write_byte(byte | 0x80)?;
+            } else {
+                break self.write_byte(byte);
+            }
+        }
+    }
+}
+
+pub struct CounterStream {
+    pub len: usize,
+}
+
+impl OutputStream for CounterStream {
+    fn write(&mut self, buf: &[u8]) -> Result<(), Error> {
+        self.len += buf.len();
+        Ok(())
+    }
+
+    fn write_byte(&mut self, _val: u8) -> Result<(), Error> {
+        self.len += 1;
+        Ok(())
+    }
+}
+
+pub struct BufferStream<'a> {
+    buf: &'a mut [u8],
+    pos: usize,
+}
+
+impl<'a> BufferStream<'a> {
+    pub fn new(buf: &'a mut [u8]) -> Self {
+        Self { buf, pos: 0 }
+    }
+
+    pub fn len(&self) -> usize {
+        self.pos
+    }
+}
+
+impl<'a> OutputStream for BufferStream<'a> {
+    fn write(&mut self, val: &[u8]) -> Result<(), Error> {
+        let pos = &mut self.pos;
```

Comment:
is this necessary or is it just to avoid writing `self.pos` repeatedly?

Reply:
It's necessary because we can't borrow `self` mutably in the closure.